internal/provider/resource_access_model.go

```diff
@@ -5,186 +5,300 @@ import (
 // access changed - apply new access
-func resourceAccessModelUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (a *accessModelResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "update")
+		return
+	}
+
+	var plan, state accessModelResourceModel
+
+	// Get the Terraform state from the request into the plan
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	var diags diag.Diagnostics
 	anyChange := false
 
 	// items that could be changed
 	var newAccess string
-	var newUsersList []string
 	var missingUserList []string
 	var addedUserList []string
 
-	var err error
+	// Check if the users has changed
+	if !plan.Users.Equal(state.Users) {
+		anyChange = true
+
+		// Get the users that are in the current state
+		stateUsers := convertRawUsers(state.Users.Elements())
 
-	if d.HasChange("users") {
+		// Get the users that are in the planned states
+		planUsers := convertRawUsers(plan.Users.Elements())
+
+		missingUserList = getMissingUsers(stateUsers, planUsers)
+		addedUserList = getAddedUsers(stateUsers, planUsers)
+	}
+
+	// Check if access has changed
+	if !plan.Access.Equal(state.Access) {
 		anyChange = true
-		oldUsers, newUsers := d.GetChange("users")
-		oldUsersInterface := oldUsers.([]interface{})
-		oldUsersList := make([]string, len(oldUsersInterface))
-		for i, v := range oldUsersInterface {
-			oldUsersList[i] = v.(string)
-		}
-		newUsersInterface := newUsers.([]interface{})
-		newUsersList = make([]string, len(newUsersInterface))
-		for i, v := range newUsersInterface {
-			newUsersList[i] = v.(string)
-		}
-		missingUserList = getMissingUsers(oldUsersList, newUsersList)
-		addedUserList = getAddedUsers(oldUsersList, newUsersList)
+		newAccess = plan.Access.ValueString()
 	}
 
 	if !anyChange {
-		return diags
+		tflog.Trace(ctx, "Update is returning without any changes.")
+		return
 	}
 
-	err = client.Models.UpdateAccessModel(juju.UpdateAccessModelInput{
-		Model:  d.Id(),
+	// TODO (cderici): The Terraform ID of this resource is leaking into the Juju package as the model identifier
+	// (despite that it's neither a uuid nor a model name). This needs to be decoupled.
+	err := a.client.Models.UpdateAccessModel(juju.UpdateAccessModelInput{
+		Model:  plan.ID.ValueString(),
 		Grant:  addedUserList,
 		Revoke: missingUserList,
 		Access: newAccess,
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access model resource, got error: %s", err))
+	}
+	// Set the plan onto the Terraform state
```

Comment:
Why isn't ID rewritten? It would have changed.

Reply:
that's a good catch, fixing 👍